forest/shared/src/cli/mod.rs

```diff
@@ -171,53 +169,76 @@ impl CliOpts {
             cfg.client.encrypt_keystore = encrypt_keystore;
         }
 
-        Ok(cfg)
+        Ok((cfg, path))
     }
 }
 
-fn find_default_config() -> Option<Config> {
-    if let Ok(config_file) = std::env::var("FOREST_CONFIG_PATH") {
-        info!(
-            "FOREST_CONFIG_PATH detected, using configuration at {}",
-            config_file
-        );
-        let path = PathBuf::from(config_file);
-        if path.exists() {
-            return read_config_or_none(path);
+pub enum ConfigPath {
+    Cli(PathBuf),
+    Env(PathBuf),
+    Project(PathBuf),
+}
+
+impl ConfigPath {
+    fn to_path_buf(&self) -> &PathBuf {
+        match self {
+            ConfigPath::Cli(path) => path,
+            ConfigPath::Env(path) => path,
+            ConfigPath::Project(path) => path,
         }
-    };
+    }
+}
 
+fn find_config_path(opts: &CliOpts) -> Option<ConfigPath> {
+    if let Some(s) = &opts.config {
+        return Some(ConfigPath::Cli(PathBuf::from(s)));
+    }
+    if let Ok(s) = std::env::var("FOREST_CONFIG_PATH") {
+        return Some(ConfigPath::Env(PathBuf::from(s)));
+    }
     if let Some(dir) = ProjectDirs::from("com", "ChainSafe", "Forest") {
-        let mut config_dir = dir.config_dir().to_path_buf();
-        config_dir.push("config.toml");
-        if config_dir.exists() {
-            info!("Found a config file at {}", config_dir.display());
-            return read_config_or_none(config_dir);
+        let mut path = dir.config_dir().to_path_buf();
+        path.push("config.toml");
+        if path.exists() {
+            return Some(ConfigPath::Project(path));
         }
     }
-
-    warn!("No configurations found, using defaults.");
-
     None
 }
 
-fn read_config_or_none(path: PathBuf) -> Option<Config> {
-    let toml = match read_file_to_string(&path) {
-        Ok(t) => t,
-        Err(e) => {
-            warn!("An error occured while reading configuration file at {}. Resorting to default configuration. Error was {}", path.display(), e);
-            return None;
+fn get_keys_flatten(value: toml::Value, keys: &mut HashSet<String>) {
```

Comment:
I see an excellent opportunity for some unit tests, especially given that there's some recursion here.